src/Database/LSMTree/Internal.hs

```diff
@@ -1278,3 +1310,115 @@ duplicate t@Table{..} = do
             tableConfig
             tableArenaManager
             content
+
+
+{-------------------------------------------------------------------------------
+   Table union
+-------------------------------------------------------------------------------}
+
+{-# SPECIALISE union :: Table IO h -> Table IO h -> IO (Table IO h) #-}
+-- | See 'Database.LSMTree.Normal.union'.
+union ::
+     (MonadMask m, MonadMVar m, MonadST m, MonadSTM m)
+  => Table m h
+  -> Table m h
+  -> m (Table m h)
+union t1 t2 = unions $ t1 :| [t2]
+
+{-# SPECIALISE unions :: NonEmpty (Table IO h) -> IO (Table IO h) #-}
+-- | See 'Database.LSMTree.Normal.unions'.
+unions ::
+     (MonadMask m, MonadMVar m, MonadST m, MonadSTM m)
+  => NonEmpty (Table m h)
+  -> m (Table m h)
+unions ts = do
+    conf <-
+      case match (fmap tableConfig ts) of
+        Left (i, j) -> throwIO $ ErrUnionsTableConfigMismatch i j
+        Right conf  -> pure conf
+
+    sesh <-
+      matchSessions ts >>= \case
+        Left (i, j) -> throwIO $ ErrUnionsSessionMismatch i j
+        Right sesh -> pure sesh
+
+    traceWith (sessionTracer sesh) $ TraceUnions (NE.map tableId ts)
```

Comment:
I think we generally trace before checking things like that session and config match. Then if the check fails, the log ends with the operation that failed.

Reply:
Problem is that we have no tracer until after we've checked the sessions, so before we've checked the sessions we can't do any tracing yet. We could take the first tracer we find in the table, but then there is the risk that they belong to different sessions and then it's not guaranteed that all tables have the same tracer. What I've done is change the order of checks and tracing: check sessions, trace, check table configs. That should make the trace appear earlier than it did previously

Reply:
Alternatively: we make the user pass in a session and use that one's tracer, and then check that the passed in session is the same as the tables' sessions. But I think it's also fine how we do it now